oci/build-layer.py

```diff
@@ -0,0 +1,299 @@
+        else:
+            return self.mode
+
+
+# Support symbolic mode changes
+class ModeChange:
+    change_regex = re.compile('(?P<op>[-+=])(?P<change>[ugo]|[rwxXst]*)')
+
+    def __init__(self, change):
+        self.change = change
+        self.cache = dict()
+
+    def apply(self, mode, is_dir=False):
+        key = (mode, is_dir)
+        cached = self.cache.get(key)
+        if cached != None:
+            return cached
+
+        mode_parts = self.change.split(',')
+
+        new_mode = mode
+        for mode_part in mode_parts:
+            index = 0
+            mask = 0
+            for c in mode_part:
+                if c == 'u':
+                    mask |= 0o700
+                elif c == 'g':
+                    mask |= 0o070
+                elif c == 'o':
+                    mask |= 0o007
+                elif c == 'a':
+                    mask |= 0o777
+                else:
+                    break
+                index += 1
+
+            if index == 0:
+                mask = 0o777
+
+            while m := self.change_regex.match(mode_part, index):
+                index = m.end()
+                op = m.group('op')
+                change_str = m.group('change')
+
+                change = 0
+                for c in change_str:
+                    if c == 'r':
+                        change |= 0o444
+                    elif c == 'w':
+                        change |= 0o222
+                    elif c == 'x':
+                        change |= 0o111
+                    elif c == 'X':
+                        if is_dir or new_mode & 0o111 > 0:
+                            change |= 0o111
+                    elif c == 's':
+                        mask |= 0o7000
+                        change |= 0o4000 | 0o2000
+                    elif c == 't':
+                        mask |= 0o7000
+                        change |= 0o1000
+                    elif c == 'u':
+                        umode = new_mode & 0o700
+                        change = umode | umode >> 3 | umode >> 6
+                    elif c == 'g':
+                        gmode = new_mode & 0o070
+                        change = gmode << 3 | gmode | gmode >> 3
+                    elif c == 'o':
+                        omode = new_mode & 0o007
+                        change = omode << 6 | omode << 3 | omode
+
+                if op == '+':
+                    new_mode = new_mode | (mask & change)
+                elif op == '-':
+                    new_mode = new_mode & ~(mask & change)
+                else:
+                    new_mode = (new_mode & ~mask) | (mask & change)
+
+            if index != len(mode_part):
+                raise Exception('Invalid mode string: {}'.format(mode_part))
+
+        self.cache[key] = new_mode
+        return new_mode
+
+
+# This class assists with avoiding duplicate entries when copying in
+# directories into the layer. OCI layers disallow having duplicate entries in
+# the tar.
```

Comment:
This was by far the most time-consuming part of making this. The failure mode is also non-obvious, doing a `skopeo copy` to docker "succeeds" but the image isn't added. Doing a `skopeo copy` to a docker archive and loading that does provide the "duplicate entries aren't allowed error".

`dockerTools` simply just shoves everything onto the filesystem and `tar`s it up which solves the duplicate problem but we can't do that because we want to avoid spinning up a VM so our files must be in the tar (otherwise the ownership information gets lost).

It's very easy to add duplicate entries especially when copying multiple directories is involved. I initially called out to `tar` for this but didn't have enough control to prevent duplicates so I had to do it all here with a lot of futz-ing.